src/pages/resultsView/fusion/ResultViewFusionMapperStore.ts

```diff
@@ -0,0 +1,212 @@
+
+export const FUS_GENE_COLORSCALE = ['#097AFF', '#3B9CFF', '#28a9ff', '#84B7DF', '#92C8E8', '#68ACEA', '#4eb7d9'];
+export const REF_GENE_COLORSCALE = ['#88C542', '#99c556', '#a0cf31', '#7ba237', '#47cb4a', '#008828', '#005623'];
+
+export class ResultViewFusionMapperStore {
+
+    /**
+     * @param {Gene} gene
+     * @param {MobxPromise<{[p: string]: CancerStudy}>} studyIdToStudy
+     * @param {MobxPromise<{[p: string]: MolecularProfile}>} molecularProfileIdToMolecularProfile
+     * @param {MobxPromise<SampleIdentifier[]>} samples
+     * @param {StructuralVariant[]} fusions
+     */
+    constructor(public gene: Gene,
+                public studyIdToStudy: MobxPromise<{ [studyId: string]: CancerStudy }>,
+                public molecularProfileIdToMolecularProfile: MobxPromise<{ [molecularProfileId: string]: MolecularProfile }>,
+                public samples: MobxPromise<SampleIdentifier[]>,
+                public fusions: StructuralVariant[]) {
+        labelMobxPromises(this);
+    }
+
+
+    /**
+     * Returns fusion data with label added as new attribute composed of site1 and site2 genes.
+     * @type {MobxPromiseUnionTypeWithDefault<StructuralVariantExt[]>}
+     */
+    readonly fusionData = remoteData({
+        invoke: async () => {
+            return (<StructuralVariantExt[]> this.fusions || []).map(f => {
+                // generate fusion label made of site1 and site2 genes
+                f.label = `${f.site1HugoSymbol}-${f.site2HugoSymbol}`;
+                return f;
+            });
+        }
+    }, []);
+
+    /**
+     * Apply color scale to the transcript
+     * @param {EnsemblTranscriptExt} d
+     * @param {boolean} isRefGene
+     * @param {EnsemblTranscriptExt[]} tempArr
+     */
+    private applyColor(d: EnsemblTranscriptExt, isRefGene: boolean, tempArr: EnsemblTranscriptExt[]) {
+
+        const attributeName: string = isRefGene ? 'transcriptId' : 'geneId';
+        const colorScale = isRefGene ? REF_GENE_COLORSCALE : FUS_GENE_COLORSCALE;
+
+        // Repeat color when length more than color scale array
+        const colorScaleIdx = tempArr.length % colorScale.length;
+        d.fillColor = colorScale[colorScaleIdx];
+
+        // collect in a temp array
+        if (tempArr.filter(t => t[attributeName] === d[attributeName]).length === 0) {
+            tempArr.push(d);
+        }
+    }
+
+    /**
+     * Returns ensembl transcripts
+     * @type {MobxPromiseUnionTypeWithDefault<Array<EnsemblTranscript>>}
+     */
+    readonly ensemblTranscripts = remoteData({
+        await: () => [
+            this.fusionData,
+        ],
+        invoke: async () => {
+            // Collect all transcript id's from the fusion data
+            let transcriptIds: any = [];
+            if (this.fusionData.result) {
+                this.fusionData.result.forEach((fusion: StructuralVariant) => {
+                    if (transcriptIds.indexOf(fusion.site1EnsemblTranscriptId) < 0) {
```

Comment:
in this case we should really use a set, because the repeated indexOfs gives quadratic time